src/test/java/com/alphawallet/attestation/demo/VerifyAttestation.sol

```diff
@@ -0,0 +1,793 @@
+        bytes tPoint;
+        uint256 challenge;
+    }
+
+    struct OctetString {
+        uint decodeIndex;
+        bytes byteCode;
+    }
+
+    //Payable variant of the attestation function for testing
+    function decodeAttestationPayable(bytes memory attestation) public returns(bool)
+    {
+        if (decodeAttestation(attestation))
+        {
+            callCount++;
+            return true;
+        }
+        else
+        {
+            return false;
+        }
+    }
+    
+    //Payable variant of the attestation function for testing
+    function verifyEqualityProofPayable(bytes memory com1, bytes memory com2, bytes memory proof) public returns(bool)
+    {
+        if (verifyEqualityProof(com1, com2, proof))
+        {
+            callCount++;
+            return true;
+        }
+        else
+        {
+            return false;
+        }
+    }
+    
+    function verifyEqualityProof(bytes memory com1, bytes memory com2, bytes memory proof) public view returns(bool)
+    {
+        Length memory len;
+        OctetString memory octet;
+        Exponent memory pok;
+        
+        len = decodeLength(proof, 1);
+        octet.decodeIndex = len.decodeIndex;
+        
+        octet = decodeOctetString(proof, octet.decodeIndex);
+        pok.base = octet.byteCode;
+        octet = decodeOctetString(proof, octet.decodeIndex);
+        pok.riddle = octet.byteCode;
+        octet = decodeOctetString(proof, octet.decodeIndex);
+        pok.challenge = bytesToUint(octet.byteCode);
+        octet = decodeOctetString(proof, octet.decodeIndex);
+        pok.tPoint = octet.byteCode;
+        
+        uint256[2] memory lhs;
+        uint256[2] memory rhs;
+        uint256[2] memory base;
+        (lhs[0], lhs[1]) = extractXYFromPoint(com1);
+        (rhs[0], rhs[1]) = extractXYFromPoint(com2);
+        (base[0], base[1]) = extractXYFromPoint(pok.base);
+        
+        //ecVals[2] = rhs[0];
+        //ecVals[3] = rhs[1];
+        
+        rhs = ecInv(rhs);
+        
+        //ecVals[4] = rhs[0];
+        //ecVals[5] = rhs[1];
+        
+        uint256[2] memory riddle = ecAdd(lhs, rhs);
+        
+        uint256[2] memory riddlePoint;
+        (riddlePoint[0], riddlePoint[1]) = extractXYFromPoint(pok.riddle);
+        
+        //ecVals[0] = riddlePoint[0];
+        //ecVals[1] = riddlePoint[1];
+        
+        //ecVals[2] = riddle[0];
+        //ecVals[3] = riddle[1];
+        
+        //Verify proof matches the commitments
+        if (!ecEquals(riddlePoint, riddle))
+        {
+            revert();
+        }
+        
+        //Ensure base is correct
+        if (!ecEquals(base, H))
+        {
+            revert();
+        }
+        
+        //Note: if needed we can hand optimise this, but the gain is small
+        //bytes memory cArray = abi.encodePacked(GPoint, pok.base, comPoint1, comPoint2, pok.tPoint); //
+        bytes memory cArray = concat5(GPoint, pok.base, com1, com2, pok.tPoint); 
```

Comment:
We strictly don't need the G point. It can be removed and will be as part of issue #72.
It is also sufficient to only hash the X coordinate and a bit describing the sign of the y coordinate. Although this would require the computation of the sign which requires a comparison to curveOrder. So I am not sure if that is better?